### PR TITLE
Memoize

### PR DESCRIPTION
**Pull request by sp3ber**
Adds memoize functions and tests for it.
(Benchmars will be later)

### CODE REVIEW

File: src/function/memoize/with.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+        const fnCallResult = fn(...args);
+
+        cache.set(cacheKey, fnCallResult);
+
+        return fnCallResult;
+    };
+});
+
+export const createCacheFrom = (obj = {}) => {
```
Comment (osorokotyaga) [inline src/function/memoize/with.js]: better use Object.create(null)

File: src/function/memoize/__tests__/deep.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        });
+        it('returns not cached value', () => {
+            const { memoizedFn, fn } = prepare();
+
+            expect(memoizedFn(1, 2)).not.toBe(memoizedFn(1, 2, 3));
+            expect(fn).toHaveBeenCalledTimes(2);
+        });
+    });
+
+    describe('with multple non-primitive arguments', () => {
```
Comment (osorokotyaga) [inline src/function/memoize/__tests__/deep.js]: test description says its for multiple arguments but actually testing only for single argument

File: src/function/memoize/deepSizeLimit.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    fn
+  )
+);
+
+function createSizedCache(maxSize) {
+    const obj = {};
+    const cache = createCacheFrom(obj);
+    const originalSetFn = cache.set;
+
+    cache.set = (...args) => {
```
Comment (osorokotyaga) [inline src/function/memoize/deepSizeLimit.js]: shouldn't it be easier to just use Set? as we anyway added Map already and it looks like it supported by browsers in README

File: src/function/memoize/shallowSingle.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * @returns {Function}
+ * @example
+ *     const addFlag = obj => ({...obj, flag: true });
+ *     const memoize = memoizeWeak(addFlag));
+ *     const test = {test: 1};
+ *     memoize(test); // from addFlag call
+ *     memoize({test: 1}); // from addFlag call
+ *     memoize(test); // from cache
+ */
+export default memoizeWith(() => new Map(), identity);
```
Comment (osorokotyaga) [inline src/function/memoize/shallowSingle.js]: it isn't a shallowSingle as it stores stores many entries and uses SameValueZero check not a shallowEqual

Comment (osorokotyaga) [reply src/function/memoize/shallowSingle.js]: ah, sorry, I understood why it called Single - it uses only one argument,
should we add one more memoize function to store result only of last call, how do you think?

File: src/function/memoize/with.js (hunk 1 above)
Comment (sp3ber) [reply src/function/memoize/with.js]: +

File: src/function/memoize/__tests__/deep.js (hunk 2 above)
Comment (sp3ber) [reply src/function/memoize/__tests__/deep.js]: fixed

File: src/function/memoize/deepSizeLimit.js (hunk 3 above)
Comment (sp3ber) [reply src/function/memoize/deepSizeLimit.js]: fixed to Map

File: src/function/memoize/shallowSingle.js (hunk 4 above)
Comment (sp3ber) [reply src/function/memoize/shallowSingle.js]: done

Comment (osorokotyaga) [reply src/function/memoize/shallowSingle.js]: seems like the name of file is a little confusing:
it called shallow but doenst perform shallowEqual check

**Comment by osorokotyaga** [comment]
current changes are merged to release/2.0.0, for other improvements please open new pr.

**Pull request closed without merging**